File: frontend/src/types/job-match.ts
export type JobMatchStatus = "PENDING" | "PROCESSING" | "COMPLETED" | "FAILED";

export type JobMatchGrade = "EXCELLENT" | "GOOD" | "MODERATE" | "LOW" | "VERY_LOW";

export type JobMatchRecommendationStatus =
  | "STRONGLY_RECOMMENDED"
  | "RECOMMENDED"
  | "CONSIDER"
  | "NOT_RECOMMENDED"
  | "INSUFFICIENT_DATA";

export type JobMatchFeedbackType =
  | "ACCURATE"
  | "TOO_HIGH"
  | "TOO_LOW"
  | "MISSING_STRENGTH"
  | "MISSING_RISK"
  | "OTHER";

export type JobMatchScores = {
  role: number;
  skill: number;
  experience: number;
  project: number;
  preference: number;
  risk: number;
};

export type JobMatchPublic = {
  id: number;
  user_id: number;
  job_posting_id: number;
  job_analysis_id: number;
  status: JobMatchStatus;
  total_score: number;
  grade: JobMatchGrade;
  recommendation_status: JobMatchRecommendationStatus;
  scores: JobMatchScores;
  matched_skills: Record<string, unknown>[];
  missing_skills: Record<string, unknown>[];
  matched_projects: Record<string, unknown>[];
  strengths: Record<string, unknown>[];
  gaps: Record<string, unknown>[];
  risks: Record<string, unknown>[];
  recommendation_summary: string | null;
  profile_completeness: number;
  profile_hash: string;
  job_analysis_hash: string;
  calculation_version: string;
  explanation_provider: string;
  is_outdated: boolean;
  calculated_at: string | null;
  created_at: string;
  updated_at: string;
};

export type JobMatchRunPublic = {
  id: number;
  job_match_id: number | null;
  user_id: number;
  job_posting_id: number;
  job_analysis_id: number;
  status: JobMatchStatus;
  profile_hash: string;
  job_analysis_hash: string;
  calculation_version: string;
  total_score: number | null;
  result_snapshot: Record<string, unknown> | null;
  error_code: string | null;
  error_message: string | null;
  started_at: string;
  completed_at: string | null;
  created_at: string;
};

export type JobMatchRunsData = {
  items: JobMatchRunPublic[];
  page: number;
  size: number;
  total: number;
  total_pages: number;
};

export type JobMatchFeedbackPublic = {
  id: number;
  job_match_id: number;
  user_id: number;
  feedback_type: JobMatchFeedbackType;
  rating: number | null;
  comment: string | null;
  created_at: string;
  updated_at: string;
};

export type JobMatchFeedbackListData = {
  items: JobMatchFeedbackPublic[];
};

File: backend/app/schemas/job_match.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field

from app.models.job import (
    JobMatchFeedbackType,
    JobMatchGrade,
    JobMatchRecommendationStatus,
    JobMatchStatus,
)


class JobMatchRunRequest(BaseModel):
    force: bool = False
    generate_explanation: bool = True


class JobMatchScores(BaseModel):
    role: int = Field(ge=0, le=100)
    skill: int = Field(ge=0, le=100)
    experience: int = Field(ge=0, le=100)
    project: int = Field(ge=0, le=100)
    preference: int = Field(ge=0, le=100)
    risk: int = Field(ge=0, le=100)


class JobMatchPublic(BaseModel):
    id: int
    user_id: int
    job_posting_id: int
    job_analysis_id: int
    status: JobMatchStatus
    total_score: int
    grade: JobMatchGrade
    recommendation_status: JobMatchRecommendationStatus
    scores: JobMatchScores
    matched_skills: list[dict[str, Any]]
    missing_skills: list[dict[str, Any]]
    matched_projects: list[dict[str, Any]]
    strengths: list[dict[str, Any]]
    gaps: list[dict[str, Any]]
    risks: list[dict[str, Any]]
    recommendation_summary: str | None
    profile_completeness: int
    profile_hash: str
    job_analysis_hash: str
    calculation_version: str
    explanation_provider: str
    is_outdated: bool
    calculated_at: datetime | None
    created_at: datetime
    updated_at: datetime


class JobMatchRunPublic(BaseModel):
    id: int
    job_match_id: int | None
    user_id: int
    job_posting_id: int
    job_analysis_id: int
    status: JobMatchStatus
    profile_hash: str
    job_analysis_hash: str
    calculation_version: str
    total_score: int | None
    result_snapshot: dict[str, Any] | None
    error_code: str | None
    error_message: str | None
    started_at: datetime
    completed_at: datetime | None
    created_at: datetime

    model_config = {"from_attributes": True}


class JobMatchRunsData(BaseModel):
    items: list[JobMatchRunPublic]
    page: int
    size: int
    total: int
    total_pages: int


class JobMatchDeletedData(BaseModel):
    deleted: bool


class JobMatchFeedbackCreate(BaseModel):
    feedback_type: JobMatchFeedbackType
    rating: int | None = Field(default=None, ge=1, le=5)
    comment: str | None = Field(default=None, max_length=2000)


class JobMatchFeedbackUpdate(BaseModel):
    feedback_type: JobMatchFeedbackType | None = None
    rating: int | None = Field(default=None, ge=1, le=5)
    comment: str | None = Field(default=None, max_length=2000)


class JobMatchFeedbackPublic(BaseModel):
    id: int
    job_match_id: int
    user_id: int
    feedback_type: JobMatchFeedbackType
    rating: int | None
    comment: str | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class JobMatchFeedbackListData(BaseModel):
    items: list[JobMatchFeedbackPublic]
